### ISSUE TITLE
CRIU crashes process and fails to checkpoint when process leverages Syscall User Dispatch

### ISSUE THREAD

**Issue by gmprice** [checkpoint-restore/criu#2037]
Syscall User Dispatch is a new feature for translating foreign system calls into linux system calls via a new interposition mechanism.

https://docs.kernel.org/admin-guide/syscall-user-dispatch.html

it does this on entry to the kernel, somewhat similar to SECCOMP

```
bool syscall_user_dispatch(struct pt_regs *regs)
{
... snip ...
  sd->on_dispatch = true;
  syscall_rollback(current, regs);
  trigger_sigsys(regs);
... snip ...
```

Presently, CRIU creates two issues in the compel and parasite code that causes a crash and failure to checkpoint.

1. The compel code forcibly blocks SIGSYS from being delivered.  Upon any future syscall (outside the Syscall User Dispatch exception area) will cause the SIGSYS to be delivered to CRIU, and CRIU will fail (usually looking for SIGTRAP).  CRIU either attempts to bail out, or it delivers the signal and waits for a TRAP.  The SIGSYS is then delivered to the process, but since SIGSYS is now masked, it crashes and everything fails.
2. In a test, I unblocked SIGSYS manually just to see what would happen.  CRIU progresses further, but upon the first compelled syscall, Syscall User Dispatch hooks that system call and attempts to deliver a SIGSYS.  At this point CRIU fails and bails out.

I believe the "correct" way to do deal with all of this is to try to get something akin to the SUSPEND_SECCOMP option merged into ptrace code

```
int ptrace_suspend_seccomp(pid_t pid)
{
  if (ptrace(PTRACE_SETOPTIONS, pid, NULL, PTRACE_O_SUSPEND_SECCOMP | PTRACE_O_TRACESYSGOOD) < 0) {
    pr_perror("suspending seccomp failed");
    return -1;
  }

  return 0;
}
```

A similar SUSPEND_SYSCALL_USER_DISPATCH option would allow all of this to happen cleanly.

**Comment by mihalicyn** [comment]
@gmprice thanks for reporting this!

Do you want to work on this feature?

**Comment by gmprice** [comment]
@mihalicyn I just pushed an RFC patch up to linux-kernel requesting SUSPEND_SYSCALL_USER_DISPATCH

https://lkml.org/lkml/2023/1/4/1044

If folks here want to review that, it would be step 1 to getting this working.  Other than that, I think this will end up being basically a 1-line fix (adding PTRACE_O_SUSPEND_SYSCALL_USER_DISPATCH to the above).

**Comment by gmprice** [comment]
Looks like there will be additional kernel work to make this work.  Specifically we'll also need a new prctl option to fetch syscall user dispatch settings.  Presently this cannot be done (they can only be set, not gotten).

Since syscall user dispatch is not security-relevant, we can just add a new PRCTL command to get the settings and this should be sufficient.

An additional ptrace(PTRACE_PEEK_SYSCALL_DISPATCH) might be needed unless CRIU can ensure the frozen task is not on a signal handler.  If we're dispatched to a SIGSYS handler, then the kernel value `current_task->syscall_dispatch.on_disaptch` cannot be restored to its proper value.  I believe the "correct" way to manage this is to simply ensure we're not on a signal handler.

This seems to be the case anyway since CRIU attempts to stop threads at entry to `__rt_sigreturn`

**Comment by gmprice** [comment]
@mihalicyn i think these are all the kernel extensions that will be needed to make this work

https://lkml.org/lkml/2023/1/9/757
https://lkml.org/lkml/2023/1/9/758
https://lkml.org/lkml/2023/1/9/759
https://lkml.org/lkml/2023/1/9/760

Going to start looking at the CRIU side, which i think should effectively drill down to

1. Detect Syscall User Dispatch being used (proc/status)
2. Invoke SUSPEND_SYSCALL_USER_DISPATCH
3. Parasite collects SUD configuration (per-thread)

Then on resume it restores the settings, like any other kernel resource.

**Comment by gmprice** [comment]
FYI - this was the latest iteration of the above kernel patch set for this as of a week ago

https://lore.kernel.org/lkml/20230131185802.GA7999@redhat.com/T/

Looking like this will hit in kernel 6.3, but this patch should allow prototyping of support.

**Comment by gmprice** [comment]
FYI this was upstreamed into 6.4: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=15bbeec0fe9f1e910cb6a173005a02b6b4582f3c
